### ISSUE TITLE
Error message is not understandable: Value with comptime-only type 'type' depends on runtime control flow

### ISSUE THREAD

**Issue by esimov** [ziglang/zig#15760]
### Zig Version

0.10.1

### Steps to Reproduce and Observed Output

I was playing with the `inline for` loops and created the following example:

```zig
fn getTypeLength(comptime T: type) usize {
    return @typeName(T).len;
}

test "inline for" {
    var values = [_]i32{ 1, 2, 3 };

    var sum: usize = 0;
    inline for (values) |val| {
        const T = switch (val) {
            1 => f32,
            2 => i8,
            3 => *const []u8,
            else => unreachable,
        };
        sum += getTypeLength(T);
    }
    std.debug.print("{any}\n", .{sum});
}
```
The compiler is giving me the following error, which honestly is barely or if not understandable at all. 

```bash
.\src\main.zig:91:19: error: value with comptime-only type 'type' depends on runtime control flow
        const T = switch (val) {
                  ^~~~~~
.\src\main.zig:91:27: note: runtime control flow here
        const T = switch (val) {
                          ^~~
.\src\main.zig:91:19: note: types are not available at runtime
```

Sure if I'm changing the values assignment on the first line from `var` to `const` the error will go away. 

```zig
// changing this line from...
var values = [_]i32{ 1, 2, 3 };
// to...
const values = [_]i32{ 1, 2, 3 }; // fix the error
```

### Expected Output

The error indicator should be on the `const` assignment and not on the `switch` part.

```zig
const values = [_]i32{ 1, 2, 3 }; 
^~~~~~ error here
``` 

I think many of the error messages the Zig compiler is throwing is not very descriptive and somehow confusing. They should be improved in order to be a bit more verbose and descriptive. Honestly for a Zig newbie this kind of message is head banger.

**Comment by esimov** [comment]
I see another ticket has been issued yesterday complaining about the same error here: https://github.com/ziglang/zig/issues/15745

**Comment by mlugg** [comment]
The reason the error message doesn't point to the `var values` is that it isn't *fundamentally* the problem here.

What's going on here is that standard `var`s are never comptime-known (and so cannot participate in compile-time evaluation). This means that in your code snippet, even thought you're using an inline loop, `val` is a runtime-known value; so that means `T` is also runtime-known, but (as the error says) that isn't allowed because types must be comptime-known. The error correctly points to the `switch` condition as the value which is runtime-known, but pointing to its declaration site would be quite strange (that would actually point you to the `inline for`'s capture here, but in general it could point to literally anything that introduces a runtime-known variable). Tracking the full "chain" of why a value is runtime-known isn't really feasible, and I doubt would even be very helpful in most cases.

Even if the error were able to magically point you to `values` and tell you it needs to be `const`, that wouldn't actually be helpful: you could fix your immediate problem, but at no point did you actually understand why it was incorrect to begin with. Not understanding part of the language and instead just hoping the compiler can always flag up the solution isn't a particularly tenable solution.

Aside from the source location of the main error being poor (discussed on the other issue you linked), I don't actually think this error is unnecessarily confusing. It's describing exactly what's going on: you have a value of type `type` (in this case the value is one of `f32`, `i8`, or `*const []u8`), which depends on runtime control flow (which it correctly points to as the switch on `val`). It can definitely be a bit hard to grasp for a beginner, but that's because what's going on here isn't trivial, and does require a bit of knowledge of this aspect of the language, for better or for worse. Of course, if you have a concrete idea on how to better word the error or useful extra information to give in it, feel free to suggest it!

**Comment by nektro** [comment]
same error on master

```
test.zig:10:19: error: value with comptime-only type 'type' depends on runtime control flow
        const T = switch (val) {
                  ^~~~~~
test.zig:10:27: note: runtime control flow here
        const T = switch (val) {
                          ^~~
test.zig:10:19: note: types are not available at runtime
```

**Comment by esimov** [comment]
I'm not arguing that the error is pointing to a wrong location on the code chain, what I'm telling is that the error is not very descriptive and probably would make sense to include into the error description the reason also why this error is happening or at least to give a hint what could be wrong. I understand that the value should be comptime known, what I don't like is the the message the compiler is giving: `Value with comptime-only type 'type'`. What is 'type' in this context?

**Comment by rohlem** [comment]
I agree that the error could be improved; IMO the compiler could absolutely trace all the dependencies back to the declaration with a note "value from variable declaration using `var` is always runtime-known" (and maybe even suggest `const` or `comptime var` as alternatives).

@esimov For clarification on the current error,
the particular `Value with comptime-only type 'type'` message points to the `switch` keyword; what it is actually referring to is the entire switch expression:
```zig
switch (val) {
  1 => f32,
  2 => i8,
  3 => *const []u8,
  else => unreachable,
}
```
Peer-type resolution of its cases (which are three case expressions of type `type`, and the `else` being type `noreturn`) leads to the overall `switch` expression being resolved to type `type` as well.
That's what the error is (a bit clumsily) communicating.

**Comment by firstrow** [comment]
@mlugg many thanks for explaining the root cause of the issue. spend 30 minutes trying to understand why `switch`.
but the error indeed was in `const` where `var` should have been used.

**Comment by akhildevelops** [comment]
```zig
for (content) |elem| {
        const h = switch (elem) {
            '0'...'9' => continue,
            '.' => continue,
            '\n' => continue,
            else => 1 + 1,
        };
    }
```

I too got a similar error:
```shell
day3/part1.zig:15:9: error: value with comptime-only type 'comptime_int' depends on runtime control flow
        switch (elem) {
        ^~~~~~
day3/part1.zig:14:10: note: runtime control flow here
    for (content) |elem| {
         ^~~~~~~
```

After reading the [comment](https://github.com/ziglang/zig/issues/15760#issuecomment-1553260929) from @mlugg understood the culprit is h which is a comptime_int that's conflicting with the run time loop. I simply,  added a typedef that's evaluated during runtime and the problem is solved
```zig
for (content) |elem| {
        const h:u8 = switch (elem) {
            '0'...'9' => continue,
            '.' => continue,
            '\n' => continue,
            else => 1 + 1,
        };
    }
```
